biolearn/model.py

```diff
@@ -289,9 +291,169 @@ def anti_trafo(x, adult_age=20):
+        "tissue": "Blood",
+        "source": "",
+        "output": "Cell Proportions",
+        "model": {
+            "type": "DeconvolutionModel",
+            "file": "EPIC_salas_18_reference.csv",
+            "platform": "EPIC",
+            "default_imputation": "none",
+        },
+    },
 }
 
 
+class DeconvolutionModel:
+    def __init__(self, reference_file, platform_input):
+
+        # sets reference data
+        self.reference = pd.read_csv(
+            get_data_file(reference_file), index_col=0
+        )
+
+        # sets reference platform type (e.g. "450K")
+        self.platform = platform_input
+
+    @classmethod
+    def from_definition(cls, clock_definition):
+
+        model_def = clock_definition["model"]
+
+        # calls constructor: "__init__()"
+        return cls(model_def["file"], model_def["platform"])
+
+    def predict(self, geo_data):
+
+        # This function computes estimate of cell proportions using quadratic programming
+        # inputs:
+        # meth_vector: ndarray vector of bulk methylation (length must equal number of rows in deconv_reference)
+        # deconv_reference: ndarray matrix with rows corresponding to CpG sites and columns to cell types
+        # outputs:
+        # cell_prop_estimate.value: ndarray vector of estimated cell type proportions
+        def solve_qp(meth_vector, deconv_reference):
+
+            A = deconv_reference
+            B = meth_vector
+
+            # confirm number of rows in reference equals methylation vector length
+            assert A.shape[0] == len(B)
+
+            # convert to standard form
+            num_cell_types = A.shape[1]
+
+            G_mat = np.dot(A.T, A)
+            a_vec = np.dot(A.T, B)
+
+            # sum of cell proportions much equal 1
+            equality_constraint = -np.ones(num_cell_types)
+
+            # cell proportions must be within [0, 1] interval
+            C_mat = np.vstack(
+                (
+                    equality_constraint,
+                    np.eye(num_cell_types),
+                    -np.eye(num_cell_types),
+                )
+            )
+            b_vec = np.concatenate(
+                [
+                    np.array([-1]),
+                    np.zeros(num_cell_types),
+                    -np.ones(num_cell_types),
+                ]
+            )
+
+            # Solve the quadratic programming problem
+            x, _, _, _, _, _ = quadprog.solve_qp(
+                G=G_mat, a=a_vec, C=C_mat.T, b=b_vec, meq=1
+            )
+
+            return x
+
+        # create copy of methylation matrix
+        meth = geo_data.dnam.copy()
+
+        # check if missing values present in methylation matrix
+        if any(meth.isna().values.flatten()):
+            print(
+                "WARNING: methylation data contains missing values. Rows with missing values will be removed"
+            )
+
+        # omit rows with missing values
+        meth.dropna(inplace=True)
```

Comment:
I'd be interested in discussing briefly the decision to handle missing values this way.